File: src/parser.cpp
#include "parser.hpp"
#include <nlohmann/json.hpp>
#include <cctype>
#include <cstdlib>
#include <sstream>
#include <string_view>

namespace {

std::string trim(const std::string& s) {
    const char* ws = " \t\r\n";
    const size_t start = s.find_first_not_of(ws);
    if (start == std::string::npos) return "";
    const size_t end = s.find_last_not_of(ws);
    return s.substr(start, end - start + 1);
}

std::string make_series_key(size_t index) {
    return "s" + std::to_string(index + 1);
}

bool is_key_char(char c) {
    return std::isalnum(static_cast<unsigned char>(c)) != 0 || c == '_' || c == '.' || c == '-';
}

bool is_key_start(char c) {
    return std::isalpha(static_cast<unsigned char>(c)) != 0 || c == '_';
}

std::string clean_key(std::string_view token) {
    size_t start = 0;
    while (start < token.size() && !is_key_char(token[start])) ++start;

    size_t end = token.size();
    while (end > start && !is_key_char(token[end - 1])) --end;

    if (start >= end || !is_key_start(token[start])) return "";
    return std::string(token.substr(start, end - start));
}

bool try_parse_number_strict(std::string token, double& out) {
    token = trim(token);
    if (token.empty()) return false;

    while (!token.empty()) {
        const char c = token.front();
        if (std::isdigit(static_cast<unsigned char>(c)) || c == '+' || c == '-' || c == '.') break;
        if (std::ispunct(static_cast<unsigned char>(c)) == 0) return false;
        token.erase(token.begin());
    }
    while (!token.empty()) {
        const char c = token.back();
        if (std::isdigit(static_cast<unsigned char>(c)) || c == '.') break;
        if (std::ispunct(static_cast<unsigned char>(c)) == 0) return false;
        token.pop_back();
    }
    if (token.empty()) return false;

    char* end = nullptr;
    out = std::strtod(token.c_str(), &end);
    return end != token.c_str() && end != nullptr && *end == '\0';
}

bool try_parse_number_relaxed(std::string token, double& out) {
    token = trim(token);
    if (token.empty()) return false;

    while (!token.empty()) {
        const char c = token.front();
        if (std::isdigit(static_cast<unsigned char>(c)) || c == '+' || c == '-' || c == '.') break;
        token.erase(token.begin());
    }
    while (!token.empty()) {
        const char c = token.back();
        if (std::isdigit(static_cast<unsigned char>(c)) || c == '.') break;
        token.pop_back();
    }
    if (token.empty()) return false;

    char* end = nullptr;
    out = std::strtod(token.c_str(), &end);
    return end != token.c_str() && end != nullptr && *end == '\0';
}

std::vector<std::string> split_whitespace(const std::string& s) {
    std::vector<std::string> tokens;
    std::istringstream ss(s);
    std::string token;
    while (ss >> token) tokens.push_back(token);
    return tokens;
}

std::vector<std::string> split_by_comma(const std::string& s) {
    std::vector<std::string> tokens;
    std::istringstream ss(s);
    std::string token;
    while (std::getline(ss, token, ',')) tokens.push_back(trim(token));
    return tokens;
}

void append_field(std::vector<ParsedField>& fields, const std::string& key, double value) {
    if (key.empty()) return;
    for (auto& field : fields) {
        if (field.key == key) {
            field.value = value;
            return;
        }
    }
    fields.push_back({key, value});
}

void flatten_numeric(const nlohmann::json& j, const std::string& prefix, std::vector<ParsedField>& out) {
    if (j.is_number()) {
        append_field(out, prefix, j.get<double>());
        return;
    }

    if (j.is_object()) {
        for (const auto& [key, value] : j.items()) {
            const std::string next = prefix.empty() ? key : prefix + "." + key;
            flatten_numeric(value, next, out);
        }
        return;
    }

    if (j.is_array()) {
        for (size_t i = 0; i < j.size(); ++i) {
            const std::string next = prefix.empty() ? make_series_key(i) : prefix + "." + std::to_string(i + 1);
            flatten_numeric(j[i], next, out);
        }
    }
}

ParsedRow parse_json_object_line(const std::string& line) {
    if (line.empty() || line.front() != '{') return {};

    nlohmann::json json;
    try {
        json = nlohmann::json::parse(line);
    } catch (...) {
        return {};
    }
    if (!json.is_object()) return {};

    ParsedRow row;
    row.fmt = LineFormat::JSONObject;
    flatten_numeric(json, "", row.fields);
    return row;
}

ParsedRow parse_json_array_line(const std::string& line) {
    if (line.empty() || line.front() != '[') return {};

    nlohmann::json json;
    try {
        json = nlohmann::json::parse(line);
    } catch (...) {
        return {};
    }
    if (!json.is_array()) return {};

    ParsedRow row;
    row.fmt = LineFormat::JSONArray;
    for (size_t i = 0; i < json.size(); ++i) {
        if (!json[i].is_number()) continue;
        row.fields.push_back({make_series_key(i), json[i].get<double>()});
    }
    return row;
}

ParsedRow parse_csv_line(const std::string& line) {
    if (line.find(',') == std::string::npos) return {};

    const auto tokens = split_by_comma(line);
    if (tokens.size() <= 1) return {};

    ParsedRow row;
    row.fmt = LineFormat::CSV;
    row.fields.reserve(tokens.size());
    for (size_t i = 0; i < tokens.size(); ++i) {
        double value = 0.0;
        if (!try_parse_number_strict(tokens[i], value)) return {};
        row.fields.push_back({make_series_key(i), value});
    }
    return row;
}

ParsedRow parse_whitespace_line(const std::string& line) {
    const auto tokens = split_whitespace(line);
    if (tokens.size() <= 1) return {};

    ParsedRow row;
    row.fmt = LineFormat::Whitespace;
    row.fields.reserve(tokens.size());
    for (size_t i = 0; i < tokens.size(); ++i) {
        double value = 0.0;
        if (!try_parse_number_strict(tokens[i], value)) return {};
        row.fields.push_back({make_series_key(i), value});
    }
    return row;
}

ParsedRow parse_key_value_log_line(const std::string& line) {
    std::string normalized;
    normalized.reserve(line.size());
    for (char c : line) {
        switch (c) {
            case ',':
            case ';':
            case '|':
            case '\t':
                normalized.push_back(' ');
                break;
            default:
                normalized.push_back(c);
                break;
        }
    }

    const auto tokens = split_whitespace(normalized);
    if (tokens.empty()) return {};

    ParsedRow row;
    row.fmt = LineFormat::KeyValueLog;

    for (size_t i = 0; i < tokens.size(); ++i) {
        const std::string& token = tokens[i];
        const size_t eq_pos = token.find('=');
        const size_t colon_pos = token.find(':');
        const size_t sep_pos = std::min(eq_pos, colon_pos);

        if (sep_pos != std::string::npos && sep_pos > 0 && sep_pos + 1 < token.size()) {
            const std::string key = clean_key(token.substr(0, sep_pos));
            double value = 0.0;
            if (!key.empty() && try_parse_number_relaxed(token.substr(sep_pos + 1), value)) {
                append_field(row.fields, key, value);
                continue;
            }
        }

        const std::string key = clean_key(token);
        if (key.empty() || i + 1 >= tokens.size()) continue;

        double value = 0.0;
        if (try_parse_number_strict(tokens[i + 1], value)) {
            append_field(row.fields, key, value);
            ++i;
        }
    }

    return row;
}

ParsedRow parse_single_value_line(const std::string& line) {
    double value = 0.0;
    if (!try_parse_number_strict(line, value)) return {};
    return ParsedRow{LineFormat::Single, {{"value", value}}};
}

}  // namespace

ParseResult detect_format(const std::string& line) {
    ParseResult result;
    const ParsedRow row = parse_line(line);
    if (row.empty()) return result;

    result.fmt = row.fmt;
    result.num_values = static_cast<int>(row.fields.size());
    result.keys.reserve(row.fields.size());
    for (const auto& field : row.fields) result.keys.push_back(field.key);
    return result;
}

ParsedRow parse_line(const std::string& line) {
    const std::string trimmed = trim(line);
    if (trimmed.empty()) return {};

    if (ParsedRow row = parse_json_object_line(trimmed); !row.empty()) return row;
    if (ParsedRow row = parse_json_array_line(trimmed); !row.empty()) return row;
    if (ParsedRow row = parse_csv_line(trimmed); !row.empty()) return row;
    if (ParsedRow row = parse_whitespace_line(trimmed); !row.empty()) return row;
    if (ParsedRow row = parse_key_value_log_line(trimmed); !row.empty()) return row;
    return parse_single_value_line(trimmed);
}

File: src/parser.rs
use serde_json::Value;

#[derive(Debug, Clone, PartialEq)]
pub enum LineFormat {
    Single,
    Whitespace,
    Csv,
    JsonObject,
    JsonArray,
    KeyValueLog,
}

#[derive(Debug, Clone)]
pub struct ParsedField {
    pub key:   String,
    pub value: f64,
}

#[derive(Debug, Clone)]
pub struct ParsedRow {
    pub format: LineFormat,
    pub fields: Vec<ParsedField>,
}

impl ParsedRow {
    pub fn is_empty(&self) -> bool {
        self.fields.is_empty()
    }
}

// ── helpers ──────────────────────────────────────────────────────────────────

fn trim(s: &str) -> &str {
    s.trim()
}

fn series_key(index: usize) -> String {
    format!("s{}", index + 1)
}

fn is_key_char(c: char) -> bool {
    c.is_alphanumeric() || c == '_' || c == '.' || c == '-'
}

fn is_key_start(c: char) -> bool {
    c.is_alphabetic() || c == '_'
}

fn clean_key(token: &str) -> Option<String> {
    let stripped = token.trim_matches(|c: char| !is_key_char(c));
    let stripped = stripped.trim_end_matches(|c: char| !is_key_char(c));
    if stripped.is_empty() {
        return None;
    }
    let first = stripped.chars().next().unwrap();
    if !is_key_start(first) {
        return None;
    }
    Some(stripped.to_string())
}

fn try_parse_strict(token: &str) -> Option<f64> {
    let s = token.trim();
    if s.is_empty() {
        return None;
    }
    // Strip leading punctuation (but keep +/-/.)
    let mut start = 0;
    for (i, c) in s.char_indices() {
        if c.is_ascii_digit() || c == '+' || c == '-' || c == '.' {
            start = i;
            break;
        } else if c.is_ascii_punctuation() {
            continue;
        } else {
            return None;
        }
    }
    // Strip trailing punctuation
    let s = &s[start..];
    let end = s
        .rfind(|c: char| c.is_ascii_digit() || c == '.')
        .map(|i| i + 1)
        .unwrap_or(0);
    let s = &s[..end];
    if s.is_empty() {
        return None;
    }
    s.parse::<f64>().ok().filter(|v| v.is_finite())
}

fn try_parse_relaxed(token: &str) -> Option<f64> {
    let s = token.trim();
    if s.is_empty() {
        return None;
    }
    // Drop leading non-numeric chars, keeping +/-/.
    let start = s.find(|c: char| c.is_ascii_digit() || c == '+' || c == '-' || c == '.')?;
    let s = &s[start..];
    let end = s
        .rfind(|c: char| c.is_ascii_digit() || c == '.')
        .map(|i| i + 1)
        .unwrap_or(0);
    let s = &s[..end];
    if s.is_empty() {
        return None;
    }
    s.parse::<f64>().ok().filter(|v| v.is_finite())
}

fn append_field(fields: &mut Vec<ParsedField>, key: String, value: f64) {
    if key.is_empty() {
        return;
    }
    for f in fields.iter_mut() {
        if f.key == key {
            f.value = value;
            return;
        }
    }
    fields.push(ParsedField { key, value });
}

fn flatten_numeric(j: &Value, prefix: &str, out: &mut Vec<ParsedField>) {
    match j {
        Value::Number(n) => {
            if let Some(v) = n.as_f64() {
                append_field(out, prefix.to_string(), v);
            }
        }
        Value::Object(map) => {
            for (k, v) in map {
                let next = if prefix.is_empty() {
                    k.clone()
                } else {
                    format!("{}.{}", prefix, k)
                };
                flatten_numeric(v, &next, out);
            }
        }
        Value::Array(arr) => {
            for (i, v) in arr.iter().enumerate() {
                let next = if prefix.is_empty() {
                    series_key(i)
                } else {
                    format!("{}.{}", prefix, i + 1)
                };
                flatten_numeric(v, &next, out);
            }
        }
        _ => {}
    }
}

// ── strategies ───────────────────────────────────────────────────────────────

fn parse_json_object(line: &str) -> Option<ParsedRow> {
    if !line.starts_with('{') {
        return None;
    }
    let j: Value = serde_json::from_str(line).ok()?;
    if !j.is_object() {
        return None;
    }
    let mut fields = Vec::new();
    flatten_numeric(&j, "", &mut fields);
    if fields.is_empty() {
        return None;
    }
    Some(ParsedRow { format: LineFormat::JsonObject, fields })
}

fn parse_json_array(line: &str) -> Option<ParsedRow> {
    if !line.starts_with('[') {
        return None;
    }
    let j: Value = serde_json::from_str(line).ok()?;
    let arr = j.as_array()?;
    let mut fields: Vec<ParsedField> = arr
        .iter()
        .enumerate()
        .filter_map(|(i, v)| v.as_f64().map(|n| ParsedField { key: series_key(i), value: n }))
        .collect();
    if fields.is_empty() {
        return None;
    }
    // Re-number to match C++ (s1, s2, …)
    for (i, f) in fields.iter_mut().enumerate() {
        f.key = series_key(i);
    }
    Some(ParsedRow { format: LineFormat::JsonArray, fields })
}

fn parse_csv(line: &str) -> Option<ParsedRow> {
    if !line.contains(',') {
        return None;
    }
    let tokens: Vec<&str> = line.split(',').map(|t| t.trim()).collect();
    if tokens.len() <= 1 {
        return None;
    }
    let fields: Vec<ParsedField> = tokens
        .iter()
        .enumerate()
        .map(|(i, t)| try_parse_strict(t).map(|v| ParsedField { key: series_key(i), value: v }))
        .collect::<Option<Vec<_>>>()?;
    Some(ParsedRow { format: LineFormat::Csv, fields })
}

fn parse_whitespace(line: &str) -> Option<ParsedRow> {
    let tokens: Vec<&str> = line.split_whitespace().collect();
    if tokens.len() <= 1 {
        return None;
    }
    let fields: Vec<ParsedField> = tokens
        .iter()
        .enumerate()
        .map(|(i, t)| try_parse_strict(t).map(|v| ParsedField { key: series_key(i), value: v }))
        .collect::<Option<Vec<_>>>()?;
    Some(ParsedRow { format: LineFormat::Whitespace, fields })
}

fn parse_kv_log(line: &str) -> Option<ParsedRow> {
    // Normalize separators to space
    let normalized: String = line
        .chars()
        .map(|c| match c {
            ',' | ';' | '|' | '\t' => ' ',
            other => other,
        })
        .collect();

    let tokens: Vec<&str> = normalized.split_whitespace().collect();
    if tokens.is_empty() {
        return None;
    }

    let mut fields: Vec<ParsedField> = Vec::new();
    let mut i = 0;
    while i < tokens.len() {
        let token = tokens[i];

        // Try key=value or key:value
        let sep_pos = token
            .find('=')
            .map(|p| (p, '='))
            .into_iter()
            .chain(token.find(':').map(|p| (p, ':')).into_iter())
            .min_by_key(|(p, _)| *p);

        if let Some((pos, _)) = sep_pos {
            if pos > 0 && pos + 1 < token.len() {
                let k = clean_key(&token[..pos]);
                let v = try_parse_relaxed(&token[pos + 1..]);
                if let (Some(key), Some(val)) = (k, v) {
                    append_field(&mut fields, key, val);
                    i += 1;
                    continue;
                }
            }
        }

        // Try: key followed by next-token value
        if let Some(key) = clean_key(token) {
            if i + 1 < tokens.len() {
                if let Some(val) = try_parse_strict(tokens[i + 1]) {
                    append_field(&mut fields, key, val);
                    i += 2;
                    continue;
                }
            }
        }

        i += 1;
    }

    if fields.is_empty() {
        return None;
    }
    Some(ParsedRow { format: LineFormat::KeyValueLog, fields })
}

fn parse_single(line: &str) -> Option<ParsedRow> {
    let v = try_parse_strict(line)?;
    Some(ParsedRow {
        format: LineFormat::Single,
        fields: vec![ParsedField { key: "value".into(), value: v }],
    })
}

// ── public API ───────────────────────────────────────────────────────────────

pub fn parse_line(line: &str) -> Option<ParsedRow> {
    let line = trim(line);
    if line.is_empty() {
        return None;
    }

    parse_json_object(line)
        .or_else(|| parse_json_array(line))
        .or_else(|| parse_csv(line))
        .or_else(|| parse_whitespace(line))
        .or_else(|| parse_kv_log(line))
        .or_else(|| parse_single(line))
}

/// Remove fields whose keys don't match the filter.
pub fn filter_fields(mut row: ParsedRow, filter: Option<&regex::Regex>) -> ParsedRow {
    if let Some(re) = filter {
        row.fields.retain(|f| re.is_match(&f.key));
    }
    row
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single() {
        let r = parse_line("42.5").unwrap();
        assert_eq!(r.format, LineFormat::Single);
        assert_eq!(r.fields[0].value, 42.5);
    }

    #[test]
    fn test_whitespace() {
        let r = parse_line("1 2 3").unwrap();
        assert_eq!(r.format, LineFormat::Whitespace);
        assert_eq!(r.fields.len(), 3);
    }

    #[test]
    fn test_csv() {
        let r = parse_line("1.0, 2.0, 3.0").unwrap();
        assert_eq!(r.format, LineFormat::Csv);
        assert_eq!(r.fields.len(), 3);
    }

    #[test]
    fn test_json_object() {
        let r = parse_line(r#"{"cpu": 72.3, "mem": 4.2}"#).unwrap();
        assert_eq!(r.format, LineFormat::JsonObject);
        assert_eq!(r.fields.len(), 2);
    }

    #[test]
    fn test_json_array() {
        let r = parse_line("[1, 2, 3]").unwrap();
        assert_eq!(r.format, LineFormat::JsonArray);
        assert_eq!(r.fields.len(), 3);
        assert_eq!(r.fields[0].key, "s1");
    }

    #[test]
    fn test_kv_log() {
        let r = parse_line("a=3 b:4 c 5").unwrap();
        assert_eq!(r.format, LineFormat::KeyValueLog);
        let keys: Vec<&str> = r.fields.iter().map(|f| f.key.as_str()).collect();
        assert!(keys.contains(&"a"));
        assert!(keys.contains(&"b"));
        assert!(keys.contains(&"c"));
    }

    #[test]
    fn test_empty() {
        assert!(parse_line("").is_none());
        assert!(parse_line("   ").is_none());
    }

    #[test]
    fn test_nested_json() {
        let r = parse_line(r#"{"a": {"b": 1, "c": 2}}"#).unwrap();
        let keys: Vec<&str> = r.fields.iter().map(|f| f.key.as_str()).collect();
        assert!(keys.contains(&"a.b"));
        assert!(keys.contains(&"a.c"));
    }
}
